### PR TITLE
SLSC Switch Scripting API

### PR DESCRIPTION
**Pull request by mdcerna**
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/niveristand-routing-and-faulting-custom-device/blob/master/CONTRIBUTING.md).

### What does this Pull Request accomplish?

First submission of the scripting API for the SLSC Switch Custom Device (within the Hardware/SLSC/ SLSC Chassis/Modules hierarchy).

### Why should this Pull Request be merged?

Adds the ability to use LabVIEW to script an SLSC Switch custom device into the system definition.

### What testing has been done?

VI Tests for all functionality has been added, which include the ability to manual inspect the resulting system definition file.

### CODE REVIEW

**Comment by mdcerna** [comment]
Johann and Donovan are gating the request, but I would appreciate all feedback.

File: Source/SLSC Switch/SLSC Module SLSC Switch.xml [hunk 1]
```diff
@@ -68,10 +68,10 @@
...
       </Name>
       <GUID>5e47c8de-3ec6-490b-8897-5526d4aea756</GUID>
-      <Glyph>
-        <Type>To Application Data Dir</Type>
-        <Path>System Explorer\Glyphs\default fpga category.png</Path>
-      </Glyph>
+			<Glyph>
+				<Type>To Common Doc Dir</Type>
+				<Path>SLSC Plugins\Modules\SLSC EDS\Glyphs\SLSC_Module.png</Path>
+			</Glyph>
```
Comment (rtzoeller) [inline Source/SLSC Switch/SLSC Module SLSC Switch.xml]: This file is a mix of tabs and spaces; pick one.

Comment (mdcerna) [reply Source/SLSC Switch/SLSC Module SLSC Switch.xml]: Spaces win.  I'll add a subset of the SLSC glyphs.

**Comment by mdcerna** [comment]
> The Scripting API build spec fails to build, saying there is a missing file. Looks like it can't find `Initialize SLSC Switch Custom Device.vi`.

Donovan, I checked out a new copy of this branch and was able to build the Scripting API from there.

File: Source/Scripting/Switch Scripting.lvlib [hunk 2]
```diff
@@ -40,6 +40,9 @@
 		<Item Name="Update Routing Channel Description.vi" Type="VI" URL="../Routing Channel/Update Routing Channel Description.vi"/>
 		<Item Name="Update Routing Channel Name.vi" Type="VI" URL="../Routing Channel/Update Routing Channel Name.vi"/>
 	</Item>
+	<Item Name="SLSC Switch" Type="Folder">
```
Comment (csjall) [inline Source/Scripting/Switch Scripting.lvlib]: Revert this change.

**Comment by csjall** [comment]
- [ ] As @buckd mentioned, you have a stray SLSC Switch scripting file referenced in the Routing and Faulting scripting library. Remove the non-existent file from the library.
image
- [ ] `Initialize SLSC Switch Custom device`: D.R.Y. Use `Set All SLSC Switch Properties` function instead of manually setting the properties. Ideally the property name constants should only be referenced in the low level getter/setter.
image
- [ ] `Set Endpoint Configuration`, `Get Endpoint Configuration`: N.I.H. Use `Format Into String` to convert enum to string. Use `Scan From String` to convert string to enum. LabVIEW supports this conversion and error handling.
image
- [ ] `test Find and Test Initialize SLSC Switch`: Clean up overlapping wires.
image
- [ ] `test Endpoint Configuration`: Replace parallel arrays for associated key/value pairs with an array of clusters with the key/value pairs. It makes it easier to read which key is associated with which value and prevents any accidental mismatched arrays.
image

**Comment by buckd** [comment]
There's a name collision when building the scripting API. There are 2 VIs named `Remove Custom Device.vi`. When the build finishes (after relinking the subVI in the lvlib), you'll get the `Routing and Faulting Scripting API.llb` with two folders next to it. The names for these VIs should be unique to avoid these extra directories.

**Comment by mdcerna** [comment]
> The Scripting API build spec fails to build, saying there is a missing file. Looks like it can't find `Initialize SLSC Switch Custom Device.vi`.

Fixed!

**Comment by csjall** [comment]
- [ ] We will have to decide whether we want to encode binary strings or not for our properties. Binary strings allows to store non ASCII strings in the file format. FlexLogger ran into problems with when storing special characters that corrupted the file. The workaround was to encode any string which may contain special characters as base64.
- [ ] Remove `test Source Separated` from `SLSC Switch Scripting Unit Tests`. Now that Scripting is in the SLSC Switch source folder, it will get evaluated by the `SLSC Switch Unit Tests` as we crawl the entire folder structure.
- [ ] The build failed in Routing and Faulting Custom Device. I was able to build it succesfully on my machine, so I am not sure what is up. I will kick another build and see if we can get a clean build.

**Comment by rtzoeller** [comment]
> We will have to decide whether we want to encode binary strings or not for our properties. Binary strings allows to store non ASCII strings in the file format. FlexLogger ran into problems with when storing special characters that corrupted the file. The workaround was to encode any string which may contain special characters as base64.

@csjall I'd like to keep using ASCII-encoded strings if we can, especially as we prioritize scripting and while we don't have an editable UI.

Is it feasible for us to write the strings normally, but fall back to encoding them as binary if the initial set property call fails?

VeriStand reads back binary strings and converts them silently, but I am not sure what happens if we try to write non-ASCII strings with the binary flag set to false.

**Comment by csjall** [comment]
> > We will have to decide whether we want to encode binary strings or not for our properties. Binary strings allows to store non ASCII strings in the file format. FlexLogger ran into problems with when storing special characters that corrupted the file. The workaround was to encode any string which may contain special characters as base64.
> 
> @csjall I'd like to keep using ASCII-encoded strings if we can, especially as we prioritize scripting and while we don't have an editable UI.
> 
> Is it feasible for us to write the strings normally, but fall back to encoding them as binary if the initial set property call fails?
> 
> VeriStand reads back binary strings and converts them silently, but I am not sure what happens if we try to write non-ASCII strings with the binary flag set to false.

I am most concerned about the capabilities field. This is a big JSON blob that is defined by vendors and we may have no control of what goes in there. I am less worried about the other fields like resource and topology.

**Comment by rtzoeller** [comment]
> > > We will have to decide whether we want to encode binary strings or not for our properties. Binary strings allows to store non ASCII strings in the file format. FlexLogger ran into problems with when storing special characters that corrupted the file. The workaround was to encode any string which may contain special characters as base64.
> > 
> > 
> > @csjall I'd like to keep using ASCII-encoded strings if we can, especially as we prioritize scripting and while we don't have an editable UI.
> > Is it feasible for us to write the strings normally, but fall back to encoding them as binary if the initial set property call fails?
> > VeriStand reads back binary strings and converts them silently, but I am not sure what happens if we try to write non-ASCII strings with the binary flag set to false.
> 
> I am most concerned about the capabilities field. This is a big JSON blob that is defined by vendors and we may have no control of what goes in there. I am less worried about the other fields like resource and topology.

I'm less concerned with the capabilities/switches JSON string being kept unencoded. I think if we can preserve the other fields we will be in a much better place.
